packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -216,23 +230,71 @@ export class Pipeline extends cdk.Construct implements events.IEventRuleTarget {
      *       onChildAdded type hook.
      * @override
      */
-    protected addChild(child: cdk.Construct, name: string) {
-        super.addChild(child, name);
+    protected addChild(child: cdk.Construct, name: string, props?: any) {
+        super.addChild(child, name, props);
         if (child instanceof Stage) {
-            this.appendStage(child);
+            this.appendStage(child, props ? (props as StageProps).placed : undefined);
         }
     }
 
-    private appendStage(stage: Stage) {
-        if (this.stages.find(x => x.name === stage.name)) {
+    private appendStage(stage: Stage, placement?: StagePlacement) {
+        if (this._stages.find(x => x.name === stage.name)) {
             throw new Error(`A stage with name '${stage.name}' already exists`);
         }
 
-        this.stages.push(stage);
+        const index = placement
+            ? this.calculateInsertIndexFromPlacement(placement)
+            : this.stagesLength;
+
+        this._stages.splice(index, 0, stage);
+    }
+
+    private calculateInsertIndexFromPlacement(placement: StagePlacement): number {
```

Comment:
Can we validate `stagePlacement` in the ctor so it's ensured to be correct when we get here?

Reply:
Can you elaborate how is that better? IMO, splitting validation and usage that assumes that validation has been done makes for less readable code.

Reply:
It’s a principle in contract-driven design: Ideally you want to validate as close as possible to the surface area and then the rest of the code can assume invariants are kept. It has a few benefits:

1. You only validate once: if in the future someone would add some other code path will want to use this value, they won’t have to duplicate the validation
2. Users see the error “closer” to where they made the mistake.

Reply:
> 2. Users see the error “closer” to where they made the mistake.

That's already the case here - `calculateInsertIndexFromPlacement`is called from `_addStage`, which in turn is called in the `Stage` constructor. So the user will see the error immediately when constructing the invalid Stage (we have a unit test confirming this behavior).

Since I'm fairly sure this parameter will not be used anywhere else, I'll leave this as it is right now, if it's OK with you.